HARK/ConsumptionSaving/ConsLabeledModel.py

```diff
@@ -0,0 +1,663 @@
+
+        return result
+
+    def evaluate(self, state):
+        """
+        Interpolate all data variables in the dataset.
+        """
+
+        if isinstance(state, xr.Dataset):
+            state_dict = state.coords
+        else:
+            state_dict = {}
+            for coords in self.dataset.coords.keys():
+                state_dict[coords] = state[coords]
+
+        result = self.dataset.interp(
+            state_dict,
+            kwargs={"fill_value": "extrapolate"},
+        )
+        result.attrs = self.dataset["v"].attrs
+
+        return result
+
+
+class ConsumerSolutionLabeled(MetricObject):
+    def __init__(
+        self,
+        value: ValueFuncCRRALabeled,
+        policy: xr.Dataset,
+        continuation: ValueFuncCRRALabeled,
+        attrs=None,
+    ):
+
+        if attrs is None:
+            attrs = dict()
+
+        self.value = value  # value function
+        self.policy = policy  # policy function
+        self.continuation = continuation  # continuation function
+
+        self.attrs = attrs
+
+    def distance(self, other):
+
+        value = self.value.dataset
+        other_value = other.value.dataset.interp_like(value)
+
+        return np.max(np.abs(value - other_value).to_array())
+
+
+class PerfForesightLabeledType(IndShockConsumerType):
+    def __init__(self, verbose=1, quiet=False, **kwds):
+        params = init_perfect_foresight.copy()
+        params.update(kwds)
+
+        # Initialize a basic AgentType
+        IndShockConsumerType.__init__(self, verbose=verbose, quiet=quiet, **params)
+
+        self.solve_one_period = make_one_period_oo_solver(
+            ConsPerfForesightLabeledSolver
+        )
+
+    def update_solution_terminal(self):
+
+        u = lambda c: CRRAutility(c, self.CRRA)  # utility
+        uP = lambda c: CRRAutilityP(c, self.CRRA)  # marginal utility
+
+        mNrm = xr.DataArray(
+            np.append(0.0, self.aXtraGrid),
+            name="mNrm",
+            dims=("mNrm"),
+            attrs={"long_name": "cash_on_hand"},
+        )
+        state = xr.Dataset({"mNrm": mNrm})  # only one state var in this model
+
+        # optimal decision is to consume everything in the last period
+        cNrm = xr.DataArray(
+            mNrm,
+            name="cNrm",
+            dims=state.dims,
+            coords=state.coords,
+            attrs={"long_name": "consumption"},
+        )
+
+        v = u(cNrm)
+        v.name = "v"
+        v.attrs = {"long_name": "value function"}
+
+        v_der = uP(cNrm)
+        v_der.name = "v_der"
+        v_der.attrs = {"long_name": "marginal value function"}
+
+        v_inv = cNrm.copy()
+        v_inv.name = "v_inv"
+        v_inv.attrs = {"long_name": "inverse value function"}
+
+        v_der_inv = cNrm.copy()
+        v_der_inv.name = "v_der_inv"
+        v_der_inv.attrs = {"long_name": "inverse marginal value function"}
+
+        dataset = xr.Dataset(
+            {
+                "cNrm": cNrm,
```

Comment:
I think it's an interesting choice that this is being labeled with `cNrm`, which names a variable, when it refers to a policy or decision rule that chooses the value for this variable based on preceding state variables.

I believe in HARK this was called something like `cFunc` in order to denote that it's a decision rule.

I wonder if there's a good reason to pick the more concise notation in this version of the model

Reply:
I assumed that was an artifact of differentiating between an array and a callable (function). Since xarray does both, I didn't think i needed to differentiate